### PR TITLE
@W-13729777 - merge small indexes together

### PR DESCRIPTION
**Pull request by sagikSF**

### CODE REVIEW

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 1]
```diff
@@ -948,5 +951,113 @@ public Bulker getBulker() {
...
 		return retryManager;
 	}
+
+    public void mergeIndexes() {
+		Set<String> currentAliases = getCurrentAliases();
+		currentAliases.forEach(alias -> {
+			List<String> indicesToBeMoved = getIndicesToBeMoved(alias);
+			List<List<String>> partition = Lists.partition(indicesToBeMoved, 2);
+			partition.forEach(indexes -> {
+				if (indexes.size() <= 1) {
```
Comment (janevelger) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:961]: Do we need try catch here per each run?

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 2]
```diff
@@ -948,5 +951,113 @@ public Bulker getBulker() {
...
+	private Set<String> getCurrentAliases() {
+		return getAliases(TIMBERMILL_INDEX_WILDCARD).keySet().stream().map(index -> index.replaceAll(INDEX_DELIMITER + "\\d+$", "")).collect(Collectors.toSet());
+	}
+
+    private List<String> getIndicesToBeMoved(String alais) {
+        try {
+			Set<String> indexNotToMerge = getAliases(alais + "*").keySet();
+            Request request = new Request("GET", "/_cat/indices/" + alais + "*?format=json");
+            Response response = client.getLowLevelClient().performRequest(request);
+            InputStream content = response.getEntity().getContent();
```
Comment (janevelger) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:980]: Should we close this input stream?

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 3]
```diff
@@ -948,5 +951,113 @@ public Bulker getBulker() {
...
+                String index = jsonObject.get("index").getAsString();
+                if (indexNotToMerge.contains(index)) {
+                    return;
+                }
+                if (isSizeLowerThen(storeSize, (long) (maxIndexSizeInGB * 0.1))) { // 10% of max index size
+                    indicesToMove.add(index);
+                }
+            });
+
+            LOG.info("Number of indices to be moved: {}", indicesToMove.size());
```
Comment (janevelger) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:996]: Lets also print the names of the indices and their sizes that we decided to move

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 4]
```diff
@@ -948,5 +951,113 @@ public Bulker getBulker() {
...
+				moveIndexAnotherIndex(sourceIndex, destIndex);
+			});
+		});
+    }
+
+	private Set<String> getCurrentAliases() {
+		return getAliases(TIMBERMILL_INDEX_WILDCARD).keySet().stream().map(index -> index.replaceAll(INDEX_DELIMITER + "\\d+$", "")).collect(Collectors.toSet());
+	}
+
+    private List<String> getIndicesToBeMoved(String alais) {
```
Comment (janevelger) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:975]:
```suggestion
    private List<String> getIndicesToBeMoved(String alias) {
```

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 5]
```diff
@@ -948,5 +951,113 @@ public Bulker getBulker() {
...
+        } catch (Exception e) {
+            LOG.error("Failed to get indices to be moved", e);
+            return Collections.emptyList();
+        }
+    }
+
+    private boolean isSizeLowerThen(String storeSize, long maxIndexSizeInGB) {
+        String sizeUnit = storeSize.substring(storeSize.length() - 2);
+
+        if (sizeUnit.equals("kb") || sizeUnit.equals("mb")) {
```
Comment (janevelger) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:1007]: can we you equals ignore case here?
Just to be sure we support all the options.

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/CronsRunner.java [hunk 6]
```diff
@@ -122,4 +125,18 @@ private void runPartialMergingTasksCron(ElasticsearchClient es, String mergingCr
 		scheduler.scheduleJob(job, trigger);
 	}
 
+	private void runIndexMergingTasksCron(ElasticsearchClient es, RedisService redisService, String indexMergingCronExp) throws SchedulerException {
+		JobDataMap jobDataMap = new JobDataMap();
+		jobDataMap.put(CLIENT, es);
+		jobDataMap.put(REDIS_SERVICE, redisService);
+		JobDetail job = newJob(IndexMergerJob.class)
+				.withIdentity("job7", "group7").usingJobData(jobDataMap)
```
Comment (janevelger) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/CronsRunner.java:133]: Is it possible to have a more meaningful identity?

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java (hunk 2 above)
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:980]: done

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java (hunk 1 above)
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:961]: done

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java (hunk 3 above)
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:996]: done but in a different place

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java (hunk 5 above)
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:1007]: done

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 7]
```diff
@@ -948,5 +951,113 @@ public Bulker getBulker() {
...
+        boolQuery.must(rangeQuery);
+
+        reindexRequest.setSourceQuery(boolQuery);
+        reindexRequest.setTimeout(TimeValue.timeValueMinutes(10));
+
+        try {
+            BulkByScrollResponse reindex = client.reindex(reindexRequest, RequestOptions.DEFAULT);
+            LOG.info("Moved {} documents from index {} to index {}", reindex.getCreated(), sourceIndex, destIndex);
+            if (isReindexSuccess(reindex)) {
+                deleteIndex(sourceIndex);
```
[the remark this answers is not in the document]
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:1034]: yes we will reindex again but the reindex works in a way that if the doc already exists and does nothing. I have tested it.  `reindexRequest.setConflicts("proceed");` this is responsible for it

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/CronsRunner.java (hunk 6 above)
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/CronsRunner.java:133]: did the same as all the others

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/IndexMergerJob.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    @Override
+    public void execute(JobExecutionContext context) throws JobExecutionException {
+        ElasticsearchClient client = (ElasticsearchClient) context.getJobDetail().getJobDataMap().get(ElasticsearchUtil.CLIENT);
+        RedisService redisService = (RedisService) context.getJobDetail().getJobDataMap().get(ElasticsearchUtil.REDIS_SERVICE);
+        JedisLock lock;
+        String flowId = "Index Merging Job - " + UUID.randomUUID();
+        MDC.put("id", flowId);
+
+        if (redisService == null) {
+            client.mergeIndexes();
```
[the remark this answers is not in the document]
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/IndexMergerJob.java:30]: it is not it is down it means we didn't pass redisService

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 9]
```diff
@@ -942,11 +938,170 @@ long countByName(String name, String env) throws IOException {
...
+            JsonArray asJsonObject = new JsonParser().parse(json).getAsJsonArray();
+            List<String> indicesToMove = new ArrayList<>();
+            asJsonObject.forEach(jsonElement -> {
+                JsonObject jsonObject = jsonElement.getAsJsonObject();
+                String storeSize = jsonObject.get("store.size").getAsString();
+                String index = jsonObject.get("index").getAsString();
+                if (indexNotToMerge.contains(index)) {
+                    return;
+                }
+                if (isSizeLowerThen(storeSize, (long) (maxIndexSizeInGB * 0.1))) { // 10% of max index size
```
Comment (tvaingart) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:989]: probably better to extract to property (the 10%) so we will be able to further experiment with that later (and maybe increase it)

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 10]
```diff
@@ -942,11 +938,170 @@ long countByName(String name, String env) throws IOException {
...
+
+    private void moveIndexAnotherIndex(String sourceIndex, String destIndex) {
+        LOG.info("Moving index {} to {}", sourceIndex, destIndex);
+        ReindexRequest reindexRequest = new ReindexRequest();
+        reindexRequest.setConflicts("proceed");
+        reindexRequest.setSourceIndices(sourceIndex);
+        reindexRequest.setDestIndex(destIndex);
+        reindexRequest.setDestOpType("create");
+        reindexRequest.setSlices(10);
+        reindexRequest.setTimeout(TimeValue.timeValueMinutes(10));
```
Comment (tvaingart) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:1031]: better to extract to property, we might need to enlarge timeouts for big re-index operations.

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 11]
```diff
@@ -942,11 +938,170 @@ long countByName(String name, String env) throws IOException {
...
-		return bulker;
-	}
+        return bulker;
+    }
+
+    public IndexRetryManager getRetryManager() {
+        return retryManager;
+    }
+
+    public void mergeIndexes() {
```
Comment (tvaingart) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:948]: potential very rare race condition, tell me if it make sense, not sure I would highly worry about it:
1. our existing (current) index is about to over
2. we generate new index to be used soon (start of rollout)
3. merge index is running: as the new index is not actually be used yet as our current alias (rollout in progress), then this index is found as potential index for merging?
4. rollout finish (this new index is being used now as current alias), while it also potentially participated in re-index

**Review by tvaingart — APPROVED** [review]
@sagikSF approved but wrote some comment, if you wanna change

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java (hunk 11 above)
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:948]: spoke about this. The chance to have a race condition here is very slim

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java (hunk 10 above)
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:1031]: there are a lot of params as it is i don't want to add more

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 12]
```diff
@@ -942,11 +938,170 @@ long countByName(String name, String env) throws IOException {
...
+                JsonObject jsonObject = new JsonParser().parse(json).getAsJsonObject();
+                boolean isCompleted = jsonObject.get("completed").getAsBoolean();
+                if (isCompleted) {
+                    JsonObject responseObject = jsonObject.get("response").getAsJsonObject();
+                    int created = responseObject.get("created").getAsInt();
+                    int numOfFailures = responseObject.get("failures").getAsJsonArray().size();
+                    LOG.info("Reindex task completed. Created {} documents. Num of failures {}", created, numOfFailures);
+                    return numOfFailures == 0;
+                } else {
+					Thread.sleep(1000);
```
Comment (edengal) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:1075]: I think you should sleep for longer, let's say 10 secs

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java (hunk 9 above)
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java]: done

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java (hunk 12 above)
Comment (sagikSF) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java]: correct my bad, fixed

**Pull request merged**